File: ios/KakaoSdkFriend/PickerApiClientModule.swift
//
//  PickerApiClientModule.swift
//  react-native-kakao-sdk-all
//
//  Created by kakao-tam on 2022/05/06.
//

import Foundation

import KakaoSDKAuth
import KakaoSDKUser
import KakaoSDKTalk
import KakaoSDKFriend

@objc(PickerApiClientModule)
class PickerApiClientModule: NSObject {
    
    @objc(selectFriend:withResolve:withReject:)
    func selectFriend(_ param:NSDictionary,
                      resolve:@escaping RCTPromiseResolveBlock,
                      reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SelectFriendParam.self)!
            PickerApi.shared.selectFriend(params:p.params) {
                (selectedUsers: SelectedUsers?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(selectedUsers?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(selectFriendPopup:withResolve:withReject:)
    func selectFriendPopup(_ param:NSDictionary,
                           resolve:@escaping RCTPromiseResolveBlock,
                           reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SelectFriendParam.self)!
            PickerApi.shared.selectFriendPopup(params:p.params) {
                (selectedUsers: SelectedUsers?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(selectedUsers?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(selectFriends:withResolve:withReject:)
    func selectFriends(_ param:NSDictionary,
                       resolve:@escaping RCTPromiseResolveBlock,
                       reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SelectFriendParam.self)!
            PickerApi.shared.selectFriends(params:p.params) {
                (selectedUsers: SelectedUsers?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(selectedUsers?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(selectFriendsPopup:withResolve:withReject:)
    func selectFriendsPopup(_ param:NSDictionary,
                            resolve:@escaping RCTPromiseResolveBlock,
                            reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SelectFriendParam.self)!
            PickerApi.shared.selectFriendsPopup(params:p.params) {
                (selectedUsers: SelectedUsers?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(selectedUsers?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
}

File: android/src/main/java/com/kakaotam/rnsdk/friend/PickerApiClientModule.kt
package com.kakaotam.rnsdk.friend

import com.facebook.react.bridge.*
import com.kakao.sdk.friend.client.PickerClient
import com.kakao.sdk.friend.model.SelectedUsers
import com.kakaotam.rnsdk.common.PromiseTransformer
import com.kakaotam.rnsdk.common.ReactToParam
import com.kakaotam.rnsdk.common.SdkInitializer
import com.kakaotam.rnsdk.friend.model.SelectFriendParam

class PickerApiClientModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "PickerApiClientModule"
    }

    @ReactMethod
    private fun selectFriend(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SelectFriendParam::class.java)!!
        PickerClient.instance.selectFriend(
            reactContext.currentActivity!!,
            p.params
        ) f@{ selectedUsers: SelectedUsers?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(selectedUsers))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun selectFriendPopup(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SelectFriendParam::class.java)!!
        PickerClient.instance.selectFriendPopup(
            reactContext.currentActivity!!,
            p.params
        ) f@{ selectedUsers: SelectedUsers?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(selectedUsers))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun selectFriends(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SelectFriendParam::class.java)!!
        PickerClient.instance.selectFriends(
            reactContext.currentActivity!!,
            p.params
        ) f@{ selectedUsers: SelectedUsers?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(selectedUsers))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun selectFriendsPopup(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SelectFriendParam::class.java)!!
        PickerClient.instance.selectFriendsPopup(
            reactContext.currentActivity!!,
            p.params
        ) f@{ selectedUsers: SelectedUsers?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(selectedUsers))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }
}
